tlvc/arch/i86/drivers/block/directfd.c

```diff
@@ -316,128 +353,154 @@ static unsigned char seek_track = 0;
 {
-#ifndef TRP_TIMER
-    static struct timer_list select = { NULL, NULL, 0, 0, select_callback };
-#endif
-
+    if (debug) printk("sel0x%x-", current_DOR);
     if (current_drive == (current_DOR & 3)) {
+	/* Drive already selected, we're ready to go */
 	floppy_ready();
 	return;
     }
+    /* activate a different drive */
     seek = 1;
     current_track = NO_TRACK;
+    /* Remember: This is drive select, not motor on/off */
+    /* NOTE: Unless the motor is turned on, select will fail */
     current_DOR &= 0xFC;
     current_DOR |= current_drive;
     outb(current_DOR, FD_DOR);
 
 #ifndef TRP_TIMER
+    /* It is not obvious why select should take any time at all ... HS */
+    /* The select_callback calls floppy_ready() */
+    /* The callback is NEVER called unless several drives are active concurrently */
     del_timer(&select);
-    select.expires = 2;
+    select.tl_expires = jiffies + 2;
     add_timer(&select);
+    set_irq();
 #endif
 }
 
-static void motor_on_callback(unsigned long nr)
+static void motor_on_callback(int nr)
 {
+    if (debug) printk("mON,");
     running |= 0x10 << nr;
     floppy_select(nr);
 }
 
 #ifndef TRP_TIMER
 static struct timer_list motor_on_timer[4] = {
-    {NULL, NULL, 0, 0, motor_on_callback},
-    {NULL, NULL, 0, 1, motor_on_callback},
-    {NULL, NULL, 0, 2, motor_on_callback},
-    {NULL, NULL, 0, 3, motor_on_callback}
+    {NULL, 0, 0, motor_on_callback},
+    {NULL, 0, 1, motor_on_callback},
+    {NULL, 0, 2, motor_on_callback},
+    {NULL, 0, 3, motor_on_callback}
+};
+static struct timer_list motor_off_timer[4] = {
+    {NULL, 0, 0, motor_off_callback},
+    {NULL, 0, 1, motor_off_callback},
+    {NULL, 0, 2, motor_off_callback},
+    {NULL, 0, 3, motor_off_callback}
 };
+static struct timer_list fd_timeout = {NULL, 0, 0, floppy_shutdown};
 #endif
 
-static void motor_off_callback(unsigned long nr)
+static void motor_off_callback(int nr)
 {
     unsigned char mask = ~(0x10 << nr);
+
+    if (debug) printk("MOF%d", nr);
     clr_irq();
     running &= mask;
     current_DOR &= mask;
     outb(current_DOR, FD_DOR);
     set_irq();
 }
 
-#ifndef TRP_TIMER
-static struct timer_list motor_off_timer[4] = {
-    {NULL, NULL, 0, 0, motor_off_callback},
-    {NULL, NULL, 0, 1, motor_off_callback},
-    {NULL, NULL, 0, 2, motor_off_callback},
-    {NULL, NULL, 0, 3, motor_off_callback}
-};
-#endif
-
-static void floppy_on(unsigned int nr)
+/*
+ * floppy_on: turn on motor. If already running, call floppy_select
+ * otherwise start motor and set timer to continue (motor_on_callback).
+ * In the latter case, motor_on_callback will call floppy_select();
+ *
+ * DOR (Data Output Register) is /MOTD/MOTC/MOTB/MOTA/DMA/(RST)/DR1/DR0/
+ */
+static void floppy_on(int nr)
 {
-    unsigned char mask = 0x10 << nr;
+    unsigned char mask = 0x10 << nr;	/* motor on select */
 
+    if (debug) printk("flpON");
+    *fl_timeout = 0;	/* Reset BIOS motor timeout counter, neccessary on some machines */
```

Comment:
The kernel IRQ 0 hardware timer interrupt calls a saved copy of the original BIOS' interrupt vector every 5 clock ticks. The BIOS usually uses this interrupt to decrement floppy motor timeouts, amongst other things.

It would be interesting to remove that BIOS callback entirely and see whether that 1) affects anything workings of the direct FD and HD drivers, thus removing another BIOS dependency, as well as 2) eliminate the need for poking a zero in the Bios Data Area to reset the motor timeout counter.

Also: given that the declaration of `fl_timeout` is `long __far *`, using `*fl_timeout = 0` zeros not only the byte location of the BIOS motor timeout at 0x440, but also three other BDA variables at 0x441-0x443!